pkg/querier/tripperware/query_attribute_matcher.go

```diff
@@ -0,0 +1,218 @@
+				if matchAttributeForExpressionQuery(attribute, r, query, expr, now, minTime, maxTime) {
+					rejectedQueriesPerTenant.WithLabelValues(userStr).Inc()
+					return httpgrpc.Errorf(http.StatusUnprocessableEntity, queryRejectErrorMessage)
+				}
+			}
+		}
+
+		reqStats := stats.FromContext(r.Context())
+		reqStats.SetDataSelectMaxTime(maxTime)
+		reqStats.SetDataSelectMinTime(minTime)
+
+		if queryPriority := limits.QueryPriority(userStr); queryPriority.Enabled && len(queryPriority.Priorities) != 0 && query != "" {
+			for _, priority := range queryPriority.Priorities {
+				for _, attribute := range priority.QueryAttributes {
+					if matchAttributeForExpressionQuery(attribute, r, query, expr, now, minTime, maxTime) {
+						reqStats.SetPriority(priority.Priority)
+						return nil
+					}
+				}
+			}
+			reqStats.SetPriority(queryPriority.DefaultPriority)
+		}
+	}
+
+	if queryReject := limits.QueryRejection(userStr); queryReject.Enabled && isMetadataQuery {
+		for _, attribute := range queryReject.QueryAttributes {
+
+			if matchAttributeForMetadataQuery(attribute, r, now) {
+				rejectedQueriesPerTenant.WithLabelValues(userStr).Inc()
+				return httpgrpc.Errorf(http.StatusUnprocessableEntity, queryRejectErrorMessage)
+			}
+		}
+	}
+
+	return nil
+}
+
+func matchAttributeForExpressionQuery(attribute validation.QueryAttribute, r *http.Request, query string, expr parser.Expr, now time.Time, minTime, maxTime int64) bool {
+	if attribute.Regex != "" && attribute.Regex != ".*" && attribute.Regex != ".+" {
+		if attribute.CompiledRegex != nil && !attribute.CompiledRegex.MatchString(query) {
+			return false
+		}
+	}
+
+	if !isWithinTimeAttributes(attribute.TimeWindow, now, minTime, maxTime) {
+		return false
+	}
+
+	if !isWithinTimeRangeAttribute(attribute.TimeRangeLimit, maxTime-minTime) {
+		return false
+	}
+
+	if !isWithinQueryStepLimit(attribute.QueryStepLimit, r, expr) {
+		return false
+	}
+
+	if attribute.UserAgent != "" && attribute.UserAgent != r.Header.Get("User-Agent") {
+		return false
+	}
+
+	if attribute.DashboardUID != "" && attribute.DashboardUID != r.Header.Get("X-Dashboard-Uid") {
+		return false
+	}
+
+	if attribute.PanelID != "" && attribute.PanelID != r.Header.Get("X-Panel-Id") {
+		return false
+	}
+
+	return true
+}
+
+func matchAttributeForMetadataQuery(attribute validation.QueryAttribute, r *http.Request, now time.Time) bool {
+	if attribute.Regex != "" && attribute.Regex != ".*" && attribute.Regex != ".+" && attribute.CompiledRegex != nil {
+		for _, matcher := range r.Form["match[]"] {
+			if attribute.CompiledRegex.MatchString(matcher) {
+				continue
+			}
+			return false
+		}
+	}
+
+	startTime, _ := util.ParseTime(r.FormValue("start"))
+	endTime, _ := util.ParseTime(r.FormValue("end"))
+
+	if !isWithinTimeAttributes(attribute.TimeWindow, now, startTime, endTime) {
+		return false
+	}
+
+	if startTime != 0 && endTime != 0 && !isWithinTimeRangeAttribute(attribute.TimeRangeLimit, endTime-startTime) {
```

Comment:
If start time is 0, why we don't check `isWithinTimeRangeAttribute`?

Reply:
For label queries, they are optional, so I thought if they are not provided, then we should skip the check. But now I moved this check to inside a `isWithinTimeRangeAttribute` function to check after the attribute check. If this attribute is set and startTime, endTime wasn't provided than it means this query doesn't match our requirements.